### DISCUSSION TITLE
Script for purge bucket (Geeetech A20M) ? (custom GCode script available)

### DISCUSSION THREAD

**Discussion by DzzD** [supermerill/SuperSlicer#1855] [Q&A] [answered]
I have now fully switched to Super Slicer, goodbye Simplify 3D !, but there is something I have trouble to find and dont know how and where I should do this.

I am looking for a purge bucket script, so that each time the extruder need to change color it will extract 10cm of filament at the left most of the bed into a bucket.

This way no need of purge block, printing is a lot faster and less filament is lost.

It was working very well with Simplify 3D, but did not find script already made or documentation to how to do it in Super Slicer.

Any help would be grealty appreciated.

**Comment by DzzD** [comment, marked as answer]
Ok, after searching and learning a bit about custom Gcode, I managed to answer my own question, script is not yet perfect but it is working on my A20M.


**Printer Setting => Tool change G-code**
************************************************************************
```
; START : CUSTOM_TOOL_CHANGE (PURGE BUCKET BY @DzzD)
G92 E0                               ; Reset absolute extruder distance to 0 
G1 E-10 F2700                   ; Retract 10mm of old filament (45mm/s)
G1 Z{layer_z+0.5} F2700    ; Lift Z +0.5mm (45mm/s)

T[next_extruder]    ; Change tool

G1 X-10 F{travel_speed*60}                             ; Go to purge bucket (travel_speed*60*mm/s)
M104 S[temperature_[next_extruder]]             ; Prepare nozzle new temperature
;***disabled*** : M109 T[next_extruder] S[temperature_[next_extruder]]   ; Wait for nozzle new temperature (should be disabled as it take looooong time even if the temperature does not change)

G92 E0             ; Reset absolute extruder distance to 0
G1 E50 F300    ; Purge by extruding 50mm of new filament (5mm/s)

G1 Z[layer_z] F{travel_speed*60}  ; Restore original z layer height (travel_speed*60*mm/s)
G92 E0                                          ; Reset extruder absolute positioning
; END: CUSTOM_TOOL_CHANGE (BUCKET BY @DzzD)
```
************************************************************************

Unfortunatly there is no access to the current_pos_x or next_pos_x as in Simplify3D so the nozzle cannot be replaced to its original location before being lowered...

**EDIT:** you should also add a g-code line to "Printer settings->Custom G-code>Start G-code" to initialise the desired first extruder/filament :

`T[current_extruder] ; set starting tool/extruder` 


This script requiere that settings : 
Printer Settings=>Extruder 1 & 2 =>Rétraction when tool is disabled=>Minimum retraction = 0
Print Setting=>Skirt/Brim=>Skirt=>Draft Shield = checked

![image](https://user-images.githubusercontent.com/12198339/141520828-fe5bc39a-fa12-46b8-a495-619b68dad59b.png)

![image](https://user-images.githubusercontent.com/12198339/141520904-265250ff-93e0-45e7-99e4-4c524963c59d.png)



**It is still WIP, but it already works pretty nice, any help, improvement are welcome !**

**Reply by supermerill** [reply]
> Unfortunatly there is no access to the current_pos_x or next_pos_x as in Simplify3D so the nozzle cannot be replaced to its original location before being lowered...

I can add that. You open the issue?

**Reply by DzzD** [reply]
I did, hope I did well :) 

Note: for the joke my A20M mainboard broke yesterday I cant do any more test until I receive a new motherboard :/ but the script posted in this discussion is fully working, and is not related to my A20M board toast ;)

**Comment by DzzD** [comment]
Works like a charm (40mn print)

![image](https://user-images.githubusercontent.com/12198339/141534798-7d5f9060-adc7-4013-b9f9-c6c50bfccc12.png) 
![image](https://user-images.githubusercontent.com/12198339/141535320-7d467978-d441-402b-b285-8735081dae78.png)

**Comment by DzzD** [comment]
I found a little error, using two extruders require to add a g-code line in **"Printer settings=>Start G-code"**, you should add this line at the end of the script : 

`T[current_extruder] ; set starting tool/extruder`

Also without this line, single color print will always use the same extruder/filament
